Add ordering-feedback attribute to pl-order-blocks

wendyguo2002 opened the pull request:
Edit (by @SethPoulsen): what ended up getting merged in this PR is a very simplified version of the feature, that doesn't include everything outlined in this description. More of the features may be added in future PRs. 

This PR only adds one new property, 'ordering-feedback', where the instructor can declare feedback which is shown when the first wrong block is an out of order block.

-------------------------------------------------

## Description
This pull request introduces a new disorder feedback feature for the DAG grading method. This enhancement enables custom feedback messages when blocks are ordered incorrectly based on specific conditions. The update also includes error handling for feedback messages within the DAG grading framework.

## Key Features

### Definitions
- **Target-Tag:** A `target-tag` is the tag of the block that is evaluated against the `condition-tag` to determine if it’s in the correct order. The disorder feedback is associated with this tag if the condition is not met.

- **Condition-Tag:** A `condition-tag` is an attribute that specifies a reference tag in the sequence. It indicates the tag that the current block should be compared against in terms of order. For example, `condition-tag="2"` points to a block with tag `2`.

- **Condition:** The `condition` attribute defines the relative position of the current block concerning the `condition-tag`. It can be set to either `after` or `before`, dictating whether the target block should appear after or before the specified `condition-tag`.

### Example of Disorder Feedback
Below is a simple discrete math example demonstrating the disorder feedback feature. The feedback will be displayed if the students don't place block with tag `1` or tag `2` before block with tag `4`.
```html
<pl-order-blocks answers-name="test-1" source-blocks-order="random" grading-method="dag" partial-credit="lcs" solution-placement="bottom">
    <pl-answer correct="true" tag="1" depends="">Let $k, n_1, n_2$ be arbitrary positive integers.</pl-answer>
    <pl-answer correct="true" tag="2" depends="1">Assume that $k|n_1$ and $k|n_2.$</pl-answer>
    <pl-answer correct="true" tag="3" depends="2">By definition, we know $q_1,q_2$ exist such that $n_1 = k\cdot q_1$ and $n_2=k\cdot q_2.$</pl-answer>
    <pl-answer correct="true" tag="4" depends="3">From here, $n_1+n_2=k\cdot q_1 + k\cdot q_2=k\cdot(q_1+q_2).$</pl-answer>
    <pl-answer correct="true" tag="5" depends="4">Now, by definition $k|(n_1+n_2).$</pl-answer>
    <pl-feedback>
        <pl-feedback-message condition-tag="1" condition="after" target-tag="4">Make sure you define k, n1, and n2 first</pl-feedback-message>
        <pl-feedback-message condition-tag="2" condition="after" target-tag="4">Other feedback</pl-feedback-message>
    </pl-feedback>
</pl-order-blocks>
```
<img width="834" alt="Screenshot 2024-08-23 at 1 55 46 PM" src="https://github.com/user-attachments/assets/0ac21bc3-dc94-43f2-9fdd-9f124475fda5">
<img width="867" alt="Screenshot 2024-08-23 at 1 56 06 PM" src="https://github.com/user-attachments/assets/5f0c789d-dd24-4915-b018-1993404b4897">
<img width="853" alt="Screenshot 2024-08-23 at 1 56 23 PM" src="https://github.com/user-attachments/assets/d50599e9-4533-419d-be12-c409d51d8bec">


### Error Handling and Validation
- **Condition-Tag Self-Reference:** An exception is raised if a `condition-tag` and `target-tag` are the same.

- **Invalid Grading Method:** An exception is raised if disorder feedback is used outside the DAG grading method.

- **Invalid Feedback Tag:** An exception is raised if any HTML tag other than `<pl-feedback-message>` is nested inside `<pl-feedback>`.

- **Missing Attributes:** An exception is raised if any required attribute (`condition-tag`, `condition`, `target-tag`) is missing in `<pl-feedback-message>`.

- **Invalid Condition Value:** An exception is raised if the `condition` attribute is not set to either "after" or "before".

wendyguo2002 commented:
I have reviewed and hereby sign the CLA

nwalters512 commented on apps/prairielearn/elements/pl-order-blocks/dag_checker.py:
```diff
@@ -55,19 +55,22 @@ def solve_dag(
-    return len(submission)
-
+    if not disordered:
+        return [-1]
```
Can you share why you're returning this instead of an empty list?

nwalters512 commented on apps/prairielearn/elements/pl-order-blocks/pl-order-blocks.mustache:
```diff
@@ -95,54 +76,78 @@ $(function() {
+                {{#disorder_feedback}}
+                <div>
+                    <span class="pl-order-blocks-feedback-icon"><i class="fa fa-question-circle" aria-hidden="true"></i></span>
+                    This block is out of order. Hint: {{disorder_feedback}}
```
I'm unsure about prefixing the feedback with this fixed text. Is there a reason we wouldn't just display the instructor-provided feedback verbatim?

nwalters512 commented on apps/prairielearn/elements/pl-order-blocks/pl-order-blocks.py:
```diff
@@ -848,12 +858,19 @@ def grade(element_html: str, data: pl.QuestionData) -> None:
+        result = grade_dag(
             submission, depends_graph, group_belonging
         )
+        num_initial_correct = result[0]
+        true_answer_length = result[1]
+        disordered_lines = result[2]
```
Is there a reason you used this instead of tuple unpacking?

nwalters512 commented on apps/prairielearn/elements/pl-order-blocks/pl-order-blocks.py:
```diff
@@ -898,12 +917,12 @@ def grade(element_html: str, data: pl.QuestionData) -> None:
                 )
             )
-            feedback = construct_feedback(
+            feedback += construct_feedback(
```
Help me understand why this was changed?

wendyguo2002 replied:
I forgot to delete "+" during the edition. I contacted Prof.Poulsen last week, and I know he won't be able to take a look until the end of August. I have read your messages above and may make a new pull request with a few changes.

nwalters512 replied:
There's no need to open a brand new PR! You can push any changes to the same branch as part of this same PR.

wendyguo2002 replied:
Thank you for the information. I am an undergraduate student at the University of Michigan. I have shown my professor this feature, and he provided some valuable feedback. Based on his advice, I will make additional enhancements to the feature. I understand it may take some time for Proof Blocks to officially incorporate my contributions, but I am passionate about participating in this development process.

nwalters512 reviewed (commented):
A few comments. I'd prefer to have @SethPoulsen review this before it lands, but he'll be offline for a while, so thanks in advance for your patience!

SethPoulsen commented:
I am just getting to looking at this now, sorry it took me so long @wendyguo2002 !

Before reading the code carefully, I have two pieces of high level feedback. 

First, I am not sure if I like the property name `disorder-feedback`. "Disorder" is not really the exact opposite of "order" in this context (English language doesn't make any sense, I know). I would prefer something like "out-of-order-feedback" or something like that, but that is too long and I would prefer it to be shorter. Does anyone have alternative suggestions?

Second, I would prefer that the new feedback be a property of the `pl-answer` tag that defines the block, instead of being on a whole new HTML tag structure. This makes it easier for people to write the feedback right next to the line it goes to. I know that the currently proposed syntax allows for multiple different feedback messages for the same line, depending on which things were before it, but is that actually a feature that we need? I think for most cases it will be enough just to have one feedback message for a line. I would rather start by adding less features, then adding more later if people need them, rather than add a whole bunch of complexity when we don't know if anyone will actually use it.

SethPoulsen commented:
Another thing--we should have tests.

You can do this by adding new questions that use the new options to this file: https://github.com/PrairieLearn/PrairieLearn/blob/master/testCourse/questions/orderBlocks/question.html

cyrus- commented:
Regarding having the feedback be a property of the `pl-answer` tag, our thought was that there will often be cases where one answer is upstream of a number of others, e.g. when a number of variables need to be introduced before they are used, so we wanted a more flexible language of booleans for expressing that. I do agree that in simple cases, maybe we can also have the option of having the feedback inline in the pl-answer tag though.

wendyguo2002 commented:
Thank you for the feedback and for taking the time to review this feature!

For the property name, I agree that "disorder-feedback" may not be the clearest term. I'll consider alternatives like "out-of-order-feedback" or "misplaced-feedback" to convey the concept more intuitively. I’ll also aim to keep the name concise.

Regarding the suggestion to make the feedback a property of the pl-answer tag: our initial thought was to create a flexible feedback system that could accommodate complex scenarios, such as when a single answer is referenced by multiple others. For instance, in cases where multiple blocks depend on defining variables before they’re used, this setup allows for more detailed guidance based on specific ordering relationships. However, I understand the concern about complexity, and I can see value in supporting a simpler option for more straightforward use cases. My plan is to implement two versions: a basic inline feedback property for simpler needs, and a more flexible, full-feature version as initially designed. This way, users can choose based on their requirements without needing to handle the additional complexity if it’s not necessary. If you think having both versions is feasible, I will implement it as soon as possible, allowing users to choose based on their specific requirements without extra complexity when it’s not needed.

Lastly, I’ll work on adding a set of questions using this feature to testCourse as soon as possible, so we can evaluate its functionality and ensure robustness. This will also help us assess if the full-featured feedback system is indeed useful for common cases or if adjustments are needed.

echuber2 commented:
For shortening the naming, I might suggest "ordering-feedback". ("order-feedback" would be even shorter, but to me this sounds like a verb, i.e. someone might expect it's a boolean attribute to control whether the feedback itself is ordered, or something.) Perhaps it's not necessary to include some explicit negative language, like "out-of-order-feedback", "incorrect-ordering-feedback", since giving feedback might imply something is incorrect already, or maybe it's useful to comment on alternative orderings that aren't incorrect anyway. So, I would just recommend "ordering-feedback".

wendyguo2002 commented:
> Another thing--we should have tests.
> 
> You can do this by adding new questions that use the new options to this file: https://github.com/PrairieLearn/PrairieLearn/blob/master/testCourse/questions/orderBlocks/question.html

Hi Prof. Poulsen,

Apologies for the delayed update. I've merged the simplified version you suggested with the more customizable version Prof. Omar recommended(the one you commented on).

The simplified version uses check-tags to link dependent blocks and validate their logical sequence. For example, a block with a check-tag enforces its placement relative to another based on a predefined order. When this order isn't followed, the system displays the corresponding ordering-feedback.
Key Validations:
Ensures referenced check-tags exist.
Verifies logical dependencies in the dependency graph.
Confirms proper usage of required attributes (e.g., check-tag, ordering-feedback).
If you need more information about this simplified version, you can check the 8/12/2024 version of the first comment I made under this pull request. The only difference is that I changed the name from disorder-feedback to ordering-feedback.

I’ve added two example questions to questions.html and will include more soon for additional clarity. Let me know if you have any questions or suggestions!

wendyguo2002 commented:
> Another thing--we should have tests.
> 
> You can do this by adding new questions that use the new options to this file: https://github.com/PrairieLearn/PrairieLearn/blob/master/testCourse/questions/orderBlocks/question.html

Hi Prof.Poulsen, I’ve added two more examples for your review. Please let me know your thoughts when you have time—no rush at all. @SethPoulsen

SethPoulsen commented:
Thanks for your work on this @wendyguo2002. I do like `ordering-feedback`, thanks for the suggestion @echuber2. 

Also thanks for adding tests. 

For this PR, I would like to only merge the syntax that uses `ordering-feedback` as a property on the `pl-answer` tag.

I think the syntax using the new `<pl-feedback>` tag is really cool, but it adds a lot of complexity to the codebase that I don't want to ship at the same time as adding the ordering feedback feature. Having code split up into more, smaller PRs makes it easier to review and avoid bugs.

Once you have made those changes, I can do a full review of this PR.

(also make sure you have addressed all the issues highlighted by @nwalters512 )

After that , if you would still like to add the new `<pl-feedback>` tag syntax, I would like it to be done as a separate pull request, including real problem examples of problems that it is necessary for.

Thanks!

SethPoulsen commented:
Also, please merge the master branch so that there are no merge conflict notifications at the bottom.

echuber2 commented:
Just thought I'd bump this since I still had the tab open. :sweat_smile:

echuber2 commented:
Hi, I tried to edit my earlier comment rather than commenting again to waste people's time. There was no meaning to my "bump" comment other than to check on the status and send a notification, which was poor etiquette. Sorry for adding meaningless noise to the conversation. I don't mean to rush you either. I do not depend on this feature for my work but I'm sure it will be useful to others.

wendyguo2002 commented:
I’ve resolved all conflicts and kept changes to existing functions minimal.  To add the ordering_feedback feature, I introduced a new function in dag_checker called get_wrong_order_lines, which is the main part of this update. In the previous version, I had included the more complex feature (pl-feedback), but I’ve now removed all related code. I’ve double-checked to make sure nothing from it is left, but there’s a chance I might have missed a few lines. Let me know if anything still needs to be cleaned up!

SethPoulsen commented on apps/prairielearn/elements/pl-order-blocks/pl-order-blocks.py:
```diff
@@ -254,6 +256,8 @@ def prepare_tag(
                     "indent",
                     "distractor-feedback",
                     "distractor-for",
+                    "ordering-feedback",
+                    "check-tag",
```
Oh, I see now. check-tag is saying which tag to check it against, is that correct? 

I see this as another feature that could be nice to have in the future, but should be left out of the initial implementation for simplicity. I think for the first version of this feature, it should just display the feedback if the block is out of order at all, not worrying about comparing to particular other blocks.

Does that work, or do you think that is a bad idea for some reason?

wendyguo2002 replied:
Yes, you got my intention with the check-tag right. I can just remove it to keep things simpler.

SethPoulsen commented on apps/prairielearn/elements/pl-order-blocks/pl-order-blocks.py:
```diff
@@ -305,6 +326,8 @@ def prepare_tag(
             "depends": depends,  # only used with DAG grader
             "group_info": group_info,  # only used with DAG grader
             "distractor_feedback": distractor_feedback,
+            "ordering_feedback": ordering_feedback,  # only used with DAG grader
```
We should be able to use this with non-DAG grader's too right?  

Its ok if  your first implementation only supports DAG grader but we should consider adding support for the other grading modes in the future.

wendyguo2002 replied:
Right now, I only implement for DAG, but I think we can add support for the other grading modes in the future. We can see how it goes first.

SethPoulsen commented on apps/prairielearn/elements/pl-order-blocks/pl-order-blocks.py:
```diff
@@ -552,7 +578,8 @@ def render(element_html: str, data: pl.QuestionData) -> str:
                 "badge_type": attempt.get("badge_type", ""),
                 "icon": attempt.get("icon", ""),
-                "distractor_feedback": attempt.get("distractor_feedback", ""),
+                "ordering_feedback": attempt.get("ordering_feedback", ""),
+                "is_disordered": attempt.get("is_disordered", False),
```
do you need to have the `is_disordered` property here, or can you base all the logic on whether or not `ordering_feedback` exists?

wendyguo2002 replied:
I will improve this part

wendyguo2002 replied:
I have tried to improve it as you said, but I have bumped into a issue. Right now I have code like:
```
for i, attempt in enumerate(student_answer):
    if i not in misordered_blocks:
        attempt["ordering_feedback"] = ""
```
This means only misordered blocks keep their ordering_feedback, and we should be able to base all the logic on whether or not ordering_feedback exists. The problem is — once a block is correct, its feedback gets cleared, and since we’re relying on student_answer to store it, that feedback is gone. After a resubmission, we can’t recover it.

To fix this, we’d need a more stable source that stores the original feedback and lets us reapply it each time. I'm not sure what the best source would be, and I’m a little concerned about increasing memory use. Another option could be keeping a flag like is_disordered to explicitly track misordered blocks — and if we go that route, I’d be happy to rename it to something like is_misordered for clarity.

Just wanted to see if that makes sense, or if there’s a better approach I might be missing. Thanks!

SethPoulsen replied:
> After a resubmission, we can’t recover it.

I don't think this is true. After each submission, during the `parse` phase, the `ordering_feedback` is accessed from `data['correct_answers']`. The properties you put here only affect the rendering of the current submission, and have nothing to do with how future submissions are rendered. 

I think you can just make this change here: and it will solve the problem

```suggestion
                "ordering_feedback":  attempt.get("ordering_feedback", "") if attempt.get("is_disordered", False) else ""
```

wendyguo2002 replied:
I’ve made some changes and fixed a few small errors. In the updated version, I also added a ValueError to ensure that the ordering-feedback attribute can only be used on blocks where correct=true. Since we already have distractor-feedback for blocks correct=false.

SethPoulsen commented on testCourse/questions/orderBlocks/question.html:
```diff
@@ -116,3 +116,42 @@
+
+<pl-order-blocks answers-name="even-implies-odd" source-blocks-order="random" grading-method="dag" partial-credit="lcs" solution-placement="bottom">
+    <pl-answer correct="true" tag="1" depends=""  >Let $k,n$ be arbitrary positive integers.</pl-answer>
+    <pl-answer correct="true" tag="2" depends="1" check-tag = "3" ordering-feedback = "Your feedback">Assume that $k|n.$</pl-answer>
```
typically we don't put spaces between the property name and equals signs, so it should be `ordering-feedback="Your feedback"`

SethPoulsen commented on apps/prairielearn/elements/pl-order-blocks/dag_checker.py:
```diff
@@ -173,6 +172,78 @@ def grade_dag(
+                    disordered_lines.append(i)
+            else:
+                # Check if there is a valid dependency between the check_tag and the current node
+                if not transitive_closure_graph.has_edge(node["check_tag"], node["tag"]) and \
+                not transitive_closure_graph.has_edge(node["tag"], node["check_tag"]):
+                    raise Exception(
+                        f"Invalid check_tag relationship: {node['tag']} and {node['check_tag']} are in different groups or have no dependency."
+                    )
```
if we get rid of `check-tag`, we shouldn't need this check either.

wendyguo2002 replied:
I've gone through all your feedback. As you pointed out, there are some small errors, and I accidentally edited parts I shouldn’t have. We can definitely remove the check-tag to keep things simple. I’ll update the function description and ensure the word "disorder" doesn’t appear. This is my first time officially merging into a big project, and I really appreciate your patience. I’ll fix those errors and make improvements.

SethPoulsen replied:
No worries! It always takes multiple rounds of code review to get things ready for commits this big, even for more experienced developers.

SethPoulsen reviewed (changes requested):
This looks good overall, I think it will be ready to merge after a couple more rounds of minor fixes. 

You also should be editing the element documentation here: https://github.com/PrairieLearn/PrairieLearn/blob/master/docs/elements.md to add information about the new `ordering-feedback` property.

wendyguo2002 commented on apps/prairielearn/elements/pl-order-blocks/dag_checker.py:
```diff
@@ -173,6 +172,78 @@ def grade_dag(
+    depends_graph: Mapping[str, list[str]],
+    group_belonging: Mapping[str, str|None]
+) ->  list[int]:
+    """
+    Determines which blocks in the student submission are disordered based on the dependency graph and group constraints.
+
+    The function ensures:
+    1. The submission adheres to the topological order of the dependency graph.
+    2. Blocks in the same pl-block-group appear contiguously.
+    3. Relationships defined by check_tag attributes are validated and enforced.
+
+    Parameters:
+        submission (list[Mapping[str, str]]): The student-submitted block ordering.
+        depends_graph (Mapping[str, list[str]]): The dependency graph specifying the logical order between blocks.
+        group_belonging (Mapping[str, Optional[str]]): A mapping of block tags to their respective pl-block-groups.
+
+    Returns:
+        list[int]: Indices of blocks that are disordered in the submission.
```
I will edit those descriptions more carefully. Thank you for your feedback.

SethPoulsen commented:
thanks for all your hard work on this @wendyguo2002, it looks like it is close to being ready to merge. Since we are getting close, I want to do a bunch of manual testing with my next review, which I don't have time for today, but hopefully I will later this week or next week.

SethPoulsen commented on apps/prairielearn/elements/pl-order-blocks/pl-order-blocks.py:
```diff
@@ -851,7 +866,9 @@ def grade(element_html: str, data: pl.QuestionData) -> None:
         elif grading_method is GradingMethodType.DAG:
             depends_graph, group_belonging = extract_dag(true_answer_list)
-
+            misordered_blocks = get_wrong_order_blocks(submission,  depends_graph, group_belonging, tags_with_feedback)    
+            for i in misordered_blocks:
+                student_answer[i]["is_misordered"] = True
```
I don't see a need for this, it doesn't look like `get_wrong_order_blocks` should ever raise an exception. 

What case were you thinking of this handling?

SethPoulsen commented on testCourse/questions/orderBlocks/question.html:
```diff
@@ -116,3 +116,23 @@
   <pl-answer correct="true">third</pl-answer>
 </pl-order-blocks>
+
+<pl-order-blocks answers-name="even-implies-odd" source-blocks-order="random" grading-method="dag" partial-credit="lcs" solution-placement="bottom">
+    <pl-answer correct="true" tag="1" depends=""  >Let $k,n$ be arbitrary positive integers.</pl-answer>
+    <pl-answer correct="true" tag="2" depends="1" check-tag = "3" ordering-feedback = "Your feedback">Assume that $k|n.$</pl-answer>
+    <pl-answer correct="true" tag="3" depends="2" >By definition of “divides”, there exists $q$ such that $n = k\cdot x.$</pl-answer>
+    <pl-answer correct="true" tag="4" depends="3" >From here, we see that $n^2 = (k\cdot x)^2=k^2\cdot x^2$.</pl-answer>
+    <pl-answer correct="true" tag="5" depends="4" >Since we found an example of a value $x^2$ that can take the role of $q$ in the definition, $k^2|n^2$, as desired.</pl-answer>
+</pl-order-blocks>
+
+
+<pl-order-blocks answers-name="hw3_q1_c" source-blocks-order="random" grading-method="dag" feedback="first-wrong" partial-credit="lcs" solution-placement="bottom">
+  <pl-answer correct="true" tag="1" depends=""  >Let $a, b, c$ be arbitrary positive integers.</pl-answer>
+  <pl-answer correct="true" tag="2" depends="1" ordering-feedback="Your feedback">Assume that $a|b$ and $b|c.$</pl-answer>
+  <pl-answer correct="true" tag="3" depends="2" >By definition, there exists $q_1$ such that $b=a\cdot q_1$ and there exists $q_2$ such that $c = b\cdot q_2.$</pl-answer>
+  <pl-answer correct="true" tag="4" depends="3" >We can see that $c = b\cdot q_2 = (a\cdot q_1)\cdot q_2 = a \cdot (q_1\cdot q_2).$</pl-answer>
+  <pl-answer correct="true" tag="5" depends="4" >By definition $a|c.$</pl-answer>
+</pl-order-blocks>
```
@wendyguo2002 looks like the check-tag property is still here, which is making the tests fail. Can you fix that? You can verify the automated tests are working by clicking the 'test' button on the 'settings' page of this question. 

Also, can you give actual example feedback here, rather than just "Your Feedback"? Or else include other example which have more realistic example feedback?

wendyguo2002 replied:
> @wendyguo2002 looks like the check-tag property is still here, which is making the tests fail. Can you fix that? You can verify the automated tests are working by clicking the 'test' button on the 'settings' page of this question.
> 
> Also, can you give actual example feedback here, rather than just "Your Feedback"? Or else include other example which have more realistic example feedback?

I have tested this question locally. It seems fine. Let me know if there is anything else I need to take a look at!

wendyguo2002 commented on apps/prairielearn/elements/pl-order-blocks/pl-order-blocks.py:
```diff
@@ -266,12 +267,21 @@ def prepare_tag(
             html_tags, "distractor-feedback", None
         )
+        ordering_feedback = pl.get_string_attrib(
+            html_tags, "ordering-feedback", None
```
> Could you make this default (and some of the other ones here) constants at the top of the file? We're trying to be more consistent about that throughout the codebase.

I have added a constant of ORDERING_FEEDBACK_ATTR at the top of the file. If this is what you want, I can do the same thing to other variables here as well.

eliotwrobson replied:
My comment was a little unclear, the default value of `None` should be set using a constant, not the attribute name (see `is_correct` a few lines above for an example). But yes, I think it would be good to add this for the other attributes too.

wendyguo2002 replied:
That makes sense. I just updated a few default constants.

SethPoulsen commented:
a few changes I made before merging:
- The feedback types we have only allow giving feedback to students up to the first wrong line, so I made it so the ordering feedback is only shown if the misordered line is the first wrong line. This is also consistent with how ordering feedback works. If you would like to make it so all the misorderd line feedback is shown, we would need create another feedback type that show all of the misordered line feedback, since it doesn't make sense to show it when the feedback type is 'first-wrong'.
- re-arranged some of the code so that the ordering feedback is handled more similarly to the distractor feedback in the code, for consistency
- With the above two changes, I was able to get rid of the 'get_wrong_lines' function entirely (of course, it can be added back if needed for the more advanced features if you want to add them)
- made it so that ordering feedback is only present under 'first-wrong-verbose', also consistent with how the distractor feedback works


Now that this is merged, if you would like to open additional PRs with some of the more complex features that this PR had in it initially, feel free to do so! Make sure to include tests and examples of how those features are used.

thanks!

SethPoulsen commented on apps/prairielearn/elements/pl-order-blocks/pl-order-blocks.py:
```diff
@@ -419,7 +440,7 @@ def prepare_tag(
     # if the order of the blocks in the HTML is a correct solution, leave it unchanged, but if it
     # isn't we need to change it into a solution before displaying it as such
     data_copy = deepcopy(data)
-    data_copy["submitted_answers"] = {answer_name: correct_answers}
+    data_copy["submitted_answers"] = {answer_name: deepcopy(correct_answers)}
```
This change was needed because the correct answers is getting passed to the grade function as a student submission, and we don't want any modifications that happen during that to persist to the actual saved correct answers.

wendyguo2002 commented:
> a few changes I made before merging:
> 
> * The feedback types we have only allow giving feedback to students up to the first wrong line, so I made it so the ordering feedback is only shown if the misordered line is the first wrong line. This is also consistent with how ordering feedback works. If you would like to make it so all the misorderd line feedback is shown, we would need create another feedback type that show all of the misordered line feedback, since it doesn't make sense to show it when the feedback type is 'first-wrong'.
> * re-arranged some of the code so that the ordering feedback is handled more similarly to the distractor feedback in the code, for consistency
> * With the above two changes, I was able to get rid of the 'get_wrong_lines' function entirely (of course, it can be added back if needed for the more advanced features if you want to add them)
> * made it so that ordering feedback is only present under 'first-wrong-verbose', also consistent with how the distractor feedback works
> * also: got it working with the `ranking` grading mode as well!
> 
> Now that this is merged, if you would like to open additional PRs with some of the more complex features that this PR had in it initially, feel free to do so! Make sure to include tests and examples of how those features are used.
> 
> thanks!
Appreciate those changes! Thanks!
